src/asn_combinators.ml

```diff
@@ -46,6 +47,21 @@ and general_string   = string 0x1b
 and universal_string = string 0x1c
 and bmp_string       = string 0x1e
 
+let enumerated (type a) (xs : (int * a) list) =
+  let module M1 = Map.Make (struct type t = int let compare = compare end) in
+  let module M2 = Map.Make (struct type t = a   let compare = compare end) in
+  let m1 = List.fold_right (fun (i, x) -> M1.add i x) xs M1.empty
+  and m2 = List.fold_right (fun (i, x) -> M2.add x i) xs M2.empty
+  in
+  let f x = try M1.find (Z.to_int x) m1 with
+    | Not_found -> parse_error @@ "integer not in enumerated"
+    | Z.Overflow -> parse_error "enumerated: overflow"
+  and g y =
+    try Z.of_int (M2.find y m2) with
+      Not_found -> parse_error @@ "value not found in enumerated"
```

Comment:
Only parsers do `parse_error` as it can't be handled otherwise. Encoders do `Invalid_argument`.